nixos/modules/services/backup/sanoid.nix

```diff
@@ -0,0 +1,163 @@
+      description = "Whether to recursively snapshot dataset children";
+      type = types.nullOr (types.enum [ true false "zfs" ]);
+      default = null;
+    };
+
+    processChildrenOnly = mkOption {
+      description = "Whether to only snapshot child datasets if recursing";
+      type = types.nullOr types.bool;
+      default = null;
+    };
+  };
+
+  configBoolPrint = value: name: optionalString (value != null) "${name} = ${if value then "yes" else "no"}";
+  configPrint = value: name: optionalString (value != null) "${name} = ${toString value}";
+
+  configBlocks = blocks: prefix:
+    concatStringsSep "\n" (mapAttrsToList (d: v: ''
+      [${prefix}${d}]
+      ${configPrint v.hourly "hourly"}
+      ${configPrint v.daily "daily"}
+      ${configPrint v.monthly "monthly"}
+      ${configPrint v.yearly "yearly"}
+
+      ${configBoolPrint v.autoprune "autoprune"}
+      ${configBoolPrint v.autosnap "autosnap"}
+
+      ${optionalString ((v.useTemplate or null) != null) ''
+        use_template = ${concatStringsSep "," v.useTemplate}
+      ''}
+
+      ${configBoolPrint (v.recursive or null) "recursive"}
+      ${configBoolPrint (v.processChildrenOnly or null) "process_children_only"}
+      ${v.extraConfig}
+    '') blocks);
+
+  configFile = ''
+    ${configBlocks cfg.datasets ""}
+
+    ${configBlocks cfg.templates "template_"}
+  '';
+
+  configDir = pkgs.runCommand "sanoid-config" {} ''
+    mkdir -p "$out"
+    ln -s "${pkgs.writeText "sanoid.conf" configFile}" "$out/sanoid.conf"
+  '';
+
+in {
+
+    # Interface
+
+    options.services.sanoid = {
+      enable = mkEnableOption "Sanoid ZFS snapshotting service";
+
+      interval = mkOption {
+        type = types.str;
+        default = "hourly";
+        example = "daily";
+        description = ''
+          Run sanoid at this interval. The default is to run hourly.
+
+          The format is described in
+          <citerefentry><refentrytitle>systemd.time</refentrytitle>
+          <manvolnum>7</manvolnum></citerefentry>.
+        '';
+      };
+
+      datasets = mkOption {
+        type = types.attrsOf (types.submodule {
+          options = datasetOptions;
+        });
+        default = {};
+        description = "Datasets to snapshot";
+      };
+
+      templates = mkOption {
+        type = types.attrsOf (types.submodule {
+          options = commonOptions;
+        });
+        default = {};
+        description = "Templates for datasets";
+      };
+
+      extraArgs = mkOption {
+        description = "Extra arguments to pass to sanoid";
+        type = types.separatedString " ";
+        default = "";
+      };
+    };
+
+    # Implementation
+
+    config = mkIf cfg.enable {
+      systemd.services.sanoid = {
+        description = "Sanoid snapshot service";
+        serviceConfig.ExecStart = "${pkgs.sanoid}/bin/sanoid --cron --configdir=${configDir} ${cfg.extraArgs}";
+        after = [ "zfs.target" ];
+        startAt = cfg.interval;
+      };
```

Comment:
sanoid should run as its own user, not root. Also consider passing `--verbose`

Reply:
Does this need `environment.TC = "UTC";`?

Reply:
I added a `user` option, but left it as `root` by default because AFAIK sanoid cannot run as another user unless ZFS privilege delegation is configured.

I'd rather not pass `--verbose` unconditionally in case someone doesn't want it for some reason. I added it to the example for `extraArgs` though.

Reply:
Yes, you need something like `/run/booted-system/sw/bin/zfs allow sanoid snapshot,destroy,mount,send,receive ${lib.escapeShellArg pool}` to allow sanoid to act on that pool.

Reply:
I don't think this module should automatically make changes like that to users' pools.

Reply:
That feels like an arbitrary distinction, you're already making changes by making and deleting snapshots. I don't see a disadvantage in this, it only applies to the sanoid user.

Reply:
I also like the idea of using `zfs allow` for more security by default, but it's not a blocker for me.

Reply:
I'm willing to implement this, but I'm not sure what approach to take. `zfs allow` needs to run as root, but `PermissionsStartOnly` is deprecated and NixOS doesn't support the command prefixes that replace it, so I don't see a way to setup privilege delegation in `preStart`. The only other way I can see is to create another service that gets started before `sanoid.service`, but that seems hacky.

Reply:
[AFAICT the prefixes can be used](https://github.com/aanderse/nixpkgs/blob/1af592d65fb3f73c536c520e990314e1ad096258/nixos/modules/services/security/munge.nix#L53), just not with pretty abstractions like `script`/`preStart`/`preStop`/etc.

Reply:
It turns out that sanoid cannot currently run as non-root. It hardcodes the path to the cache file as `/var/cache/sanoidsnapshots.txt`, which cannot be created as an unprivileged user. This could be worked around with tmpfiles or another `ExecStartPre` command. The real problem is that it puts lock files directly in `/var/run`. In this case, sanoid needs to create and delete these files itself, so I don't see a way to work around this.

I am working on patches to fix these issues, which I will submit upstream.

Reply:
See: https://github.com/jimsalterjrs/sanoid/pull/463

Reply:
The `--run-dir` and `--cache-dir` options have been accepted upstream. There is no new release yet, so I added them as patches.